Composite Key migration for citus

sravfeyn opened the pull request:
@snopoke

sravfeyn committed:
add composite key migration

sravfeyn committed:
Merge branch 'master' into sr-citus-co

sravfeyn committed:
resolve migration order

sravfeyn commented on custom/icds_reports/migrations/0103_aggregateccsrecordcomplementaryfeedingforms_supervisor_id.py:
```diff
@@ -9,7 +9,7 @@
 
     dependencies = [
-        ('icds_reports', '0101_add_supervisor_id'),
+        ('icds_reports', '0102_drop_update_aggregate_awc_data'),
```
This is also being addressed in https://github.com/dimagi/commcare-hq/pull/23543/ since it's more urgent

sravfeyn commented:
Is this all, I feel there's more to it? I can update the agg-tests to have `IS_ON_CITUSDB_BACKEND` set to True. But without docker setup with Citus, that's not useful?

snopoke commented on custom/icds_reports/migrations/0104_citus_composite_key.py:
```diff
@@ -0,0 +1,55 @@
+        'dailyattendance',
+        'awwincentivereport'
+    ]
+    if not getattr(settings, 'IS_ON_CITUSDB_BACKEND', False):
```
This flag doesn't exist - I think you can use the `UNIT_TEST` flag for now

sravfeyn replied:
We can create one and use that? I will use `UNIT_TEST` for now.

snopoke commented on custom/icds_reports/migrations/0104_citus_composite_key.py:
```diff
@@ -0,0 +1,55 @@
+
+
+def _citus_composite_key_migration(apps, schema_editor):
+    models_to_update = [
```
was this list copied from the other branch? Please check it has all the right tables.

sravfeyn replied:
Yeah, I think missed checking two. Updated https://github.com/dimagi/commcare-hq/pull/23544/commits/51d7c8feeecb054cc21fb1e081d0ee763e53692f

snopoke commented on custom/icds_reports/migrations/0104_citus_composite_key.py:
```diff
@@ -0,0 +1,55 @@
+    ]
+
+    operations = [
+        HqRunPython(_citus_composite_key_migration),
```
This will need corresponding Django state operations otherwise Django will think the indexes haven't been added. Probably the best route would be to make the function above work for a single table and then you can have something like this:
```
operations = [
    migrations.RunSQL(get_composite_index_sql('icds-reports', 'modelname'), state_operations=....)
]
```

sravfeyn replied:
Since Django doesn't support composite key, how does an example state operation look like for a model?

snopoke replied:
it would be the state operation to add the unique index which you've added to the django models

snopoke commented on custom/icds_reports/models/aggregate.py:
```diff
@@ -862,6 +862,7 @@ class DailyAttendance(models.Model):
     class Meta:
         managed = False
         db_table = 'daily_attendance'
+        unique_together = ('supervisor_id', 'case_id')
```
The primary key here is `doc_id` not `case_id`, please check the others.

sravfeyn replied:
Updated https://github.com/dimagi/commcare-hq/pull/23544/commits/51d7c8feeecb054cc21fb1e081d0ee763e53692f

sravfeyn committed:
remove two agg tables from composite key list

sravfeyn committed:
Merge branch 'master' into sr-citus-co

sravfeyn committed:
migration dependency

sravfeyn committed:
add state_operations

sravfeyn committed:
add another agg table

sravfeyn committed:
remove prints use UNIT_TESTING flag

sravfeyn commented:
@snopoke I have updated this now

snopoke commented on custom/icds_reports/migrations/0104_citus_composite_key.py:
```diff
@@ -0,0 +1,82 @@
+    if getattr(settings, 'UNIT_TESTING', False):
+        return sql, reverse_sql
+    else:
+        return migrations.RunSQL.noop, reverse_sql
```
why would this not be a noop for the reverse as well?

emord commented on custom/icds_reports/migrations/0104_citus_composite_key.py:
```diff
@@ -0,0 +1,82 @@
+    models_to_update = [
+        'aggregatebirthpreparednesforms',
+        'aggregateccsrecorddeliveryforms',
+        'aggregateccsrecordpostnatalcareforms',
```
Has this been checked against currently existing tables with partitions? I think that constraints aren't automatically propagated to child tables. I'm not 100% sure if that differs for PRIMARY KEY, but I don't believe that it is any different.

sravfeyn replied:
In fact, it looks like it doesn't get propagated to child tables, because of postgres's limitation of inheritance on unique/primary-key constraints and index. See caveats under https://www.postgresql.org/docs/9.5/ddl-inherit.html. This means we will probably have to run similar SQL on all child tables? Is there any other better way?

sravfeyn replied:
(I had tested this manually on a local pg db)

snopoke replied:
Since this is not going to be run on the live site (only for test setup), I do not believe the propagation to the child tables is an issue. 

For any tables that remain partitioned in future converting them to use PG native partitioning will result in the primary key being propagated however it also requires that the partition column be in the primary key (I don't think I realised that before and not sure how I got around it before but seems like extra reason not to have tables partitioned).

sravfeyn committed:
better way to get db connection

emord commented on custom/icds_reports/migrations/0104_citus_composite_key.py:
```diff
@@ -0,0 +1,82 @@
+    connection = connections[db_for_read_write(model_cls)]
+    with connection.schema_editor() as schema_editor:
+        pkey_name = '{}_pkey'.format(model_cls._meta.db_table)
+        fields = list(model_cls._meta.unique_together)[0]
```
is this how it gets around not creating the indexes when you run `makemigrations`? I think this would break if someone attempted to add a `unique_together` constraint before the currently defined one. is that correct? 

it seems pretty hacky, but probably low likelihood to ever be hit. Could you add a comment about how this is working in case someone does run into in the future?

sravfeyn replied:
(if I understand the question correctly) No this isn't. Because `state_operations` are being passed, `makemigrations` won't create a new migration file. But, if someone modifies `unique_together` and does a `makemigrations`, it _will_ create the necessary migration file that creates the indices.

sravfeyn committed:
Merge branch 'master' into sr-citus-co

calellowitz committed:
fix pk migrations

sravfeyn committed:
Merge branch 'master' into sr-citus-co

snopoke committed:
Merge branch 'master' into sr-citus-co

calellowitz commented:
@sravfeyn 
Not sure if you are still struggling to get tests passing, but I think you will have to make changes similar to the ones I made in https://github.com/dimagi/commcare-hq/blob/ce/distribute-ucr/custom/icds_reports/migrations/0104_citus_composite_key.py

You need to use `RawSQLMigration` for it to run on the correct dbs at the correct times. You may just be able to apply the changes I made, although it is worth double checking to make sure they seem right to you as well

sravfeyn commented:
Thanks Cal, I have just updated the file, will see if it goes okay.

snopoke committed:
add month to the primary key for montly tables

snopoke committed:
add missing model

snopoke committed:
add state operations

snopoke committed:
no need for SQL class

snopoke commented:
@calellowitz why is the use of the `RawSqlMigrator` necessary?

snopoke committed:
Merge branch 'master' into sr-citus-co

snopoke committed:
resolve migration conflict

snopoke committed:
split commands

PostgreSQL v10+ prevents execution of ADD CONSTRAINT
commands with other ALTER TABLE subcommands

snopoke committed:
add missing supervisor IDs

snopoke committed:
Merge branch 'master' into sr-citus-co

snopoke committed:
resolve migration conflict

snopoke commented:
:white_check_mark: build passing

snopoke commented:
@calellowitz can you do a final pass on this before merging

calellowitz reviewed (approved):
Looks good to me
